src/realm/node_header.hpp

```diff
@@ -229,16 +301,631 @@ class NodeHeader {
+            // old layout
+            uint8_t wtype = static_cast<uint8_t>(enc);
+            hb[4] = (flags << 5) | (wtype << 3);
+            if (enc == Encoding::WTypBits)
+                set_width_in_header(bits_pr_elem, reinterpret_cast<char*>(header));
+            else
+                set_width_in_header(bits_pr_elem >> 3, reinterpret_cast<char*>(header));
+            set_size_in_header(num_elems, reinterpret_cast<char*>(header));
+        }
+        else if (enc == Encoding::Packed) {
+            hb[2] = 0;
+            hb[3] = static_cast<uint8_t>(bits_pr_elem);
+            hb[4] = (flags << 5) | (wtype_Extend << 3);
+            hb[5] = static_cast<uint8_t>(enc) - wtype_Extend;
+            const auto hw = reinterpret_cast<uint16_t*>(header);
+            hw[3] = static_cast<uint16_t>(num_elems);
+        }
+    }
+
+    // valid for encodings with two element sizes, but only one number of elements
+    static void init_header(char* header, Encoding enc, uint8_t flags, uint8_t bits_pr_elemA, uint8_t bits_pr_elemB,
+                            size_t num_elems)
+    {
+        std::fill(header, header + header_size, 0);
+        const auto hb = reinterpret_cast<uint8_t*>(header);
+        REALM_ASSERT_DEBUG(enc == Encoding::AofP || enc == Encoding::PofA);
+        if (enc == Encoding::AofP) {
+            hb[4] = (flags << 5) | (wtype_Extend << 3);
+            hb[5] = static_cast<int>(Encoding::AofP) - static_cast<int>(Encoding::Packed);
+        }
+        else if (enc == Encoding::PofA) {
+            hb[4] = (flags << 5) | (wtype_Extend << 3);
+            hb[5] = static_cast<int>(Encoding::PofA) - static_cast<int>(Encoding::Packed);
+        }
+        hb[2] = static_cast<uint8_t>(bits_pr_elemA);
+        hb[3] = static_cast<uint8_t>(bits_pr_elemB);
+        const auto hw = reinterpret_cast<uint16_t*>(header);
+        hw[3] = static_cast<uint16_t>(num_elems);
+    }
+
+    static void init_header(char* header, Encoding enc, uint8_t flags, uint8_t bits_pr_elemA, uint8_t bits_pr_elemB,
+                            size_t num_elemsA, size_t num_elemsB)
+    {
+        std::fill(header, header + header_size, 0);
+        const auto hb = reinterpret_cast<uint8_t*>(header);
+        REALM_ASSERT_DEBUG(enc == Encoding::Flex);
+        REALM_ASSERT_DEBUG(flags < 8);
+        hb[4] = (flags << 5) | (wtype_Extend << 3);
+        hb[5] = static_cast<int>(Encoding::Flex) - static_cast<int>(Encoding::Packed);
+        const auto hw = reinterpret_cast<uint16_t*>(header);
+        REALM_ASSERT_DEBUG(bits_pr_elemA > 0);
+        REALM_ASSERT_DEBUG(bits_pr_elemB > 0);
+        REALM_ASSERT_DEBUG(bits_pr_elemA <= 64);
+        REALM_ASSERT_DEBUG(bits_pr_elemB <= 64);
+        REALM_ASSERT_DEBUG(num_elemsA < 1024);
+        REALM_ASSERT_DEBUG(num_elemsB < 1024);
+        hw[1] = static_cast<uint16_t>(((bits_pr_elemB - 1) << 10) | num_elemsB);
+        hw[3] = static_cast<uint16_t>(((bits_pr_elemA - 1) << 10) | num_elemsA);
+    }
+
```

Comment:
The following functions should be private or in anonymous namespace.

Reply:
They cannot be private (maybe protected), because `Arrays` inherits from `Node`, which inherits from `NodeHeader`.  I can add them to the anonymous namespace (but they use a lot of internal `NodeHeader` stuff, it will make the code a bit more verbose.)